resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/DriverCommandFeatureStep.scala

```diff
@@ -0,0 +1,137 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.deploy.k8s.features
+
+import scala.collection.JavaConverters._
+
+import io.fabric8.kubernetes.api.model.{ContainerBuilder, EnvVarBuilder}
+
+import org.apache.spark.deploy.k8s._
+import org.apache.spark.deploy.k8s.Config._
+import org.apache.spark.deploy.k8s.Constants._
+import org.apache.spark.deploy.k8s.submit._
+import org.apache.spark.internal.config._
+import org.apache.spark.launcher.SparkLauncher
+import org.apache.spark.util.Utils
+
+/**
+ * Creates the driver command for running the user app, and propagates needed configuration so
+ * executors can also find the app code.
+ */
+private[spark] class DriverCommandFeatureStep(conf: KubernetesConf[KubernetesDriverSpecificConf])
+  extends KubernetesFeatureConfigStep {
+
+  private val driverConf = conf.roleSpecificConf
+
+  override def configurePod(pod: SparkPod): SparkPod = {
+    driverConf.mainAppResource match {
+      case Some(JavaMainAppResource(_)) | None =>
+        configureForJava(pod)
+
+      case Some(PythonMainAppResource(res)) =>
+        configureForPython(pod, res)
+
+      case Some(RMainAppResource(res)) =>
+        configureForR(pod, res)
+    }
+  }
+
+  override def getAdditionalPodSystemProperties(): Map[String, String] = {
+    driverConf.mainAppResource match {
+      case Some(JavaMainAppResource(res)) =>
+        additionalJavaProperties(res)
+
+      case Some(PythonMainAppResource(res)) =>
+        additionalPythonProperties(res)
+
+      case Some(RMainAppResource(res)) =>
+        additionalRProperties(res)
+
+      case None =>
+        Map.empty
```

Comment:
I believe so. The code in `KubernetesClientApplication.scala` can leave the resource as `None`, and these were even unit tests (which I just moved to a new place) to make sure that it translated to a JVM-based driver.

Reply:
Ah: using `JavaMainAppResource(NO_RESOURCE)` would also work, but it means that there would be logic to ignore that elsewhere, where the resource is added to the jars that are part of the application (in this same class).

Reply:
Given the current logic I think `JavaMainAppResource(Option[String])` would make more sense, since "no resource" translates to Java. If that's ok I can make that change - including making it not optional in `KubernetesConf`.

Reply:
Sounds great, thanks!